packages/parbox.lua

```diff
@@ -0,0 +1,544 @@
+  -- method, as we enforced the 1 in leaveHMode.
+  -- parboxTypesetter.endline = function (self)
+  --   self:leaveHmode(1)
+  --   SILE.documentState.documentClass.endPar(self)
+  -- end
+  local parboxFrame = parboxTempFrame(options)
+  parboxTypesetter:init(parboxFrame)
+  SILE.typesetter = parboxTypesetter
+
+  SILE.process(content)
+  parboxTypesetter:leaveHmode(1)
+  innerVbox = parboxTypesetter.state.outputQueue
+
+  SILE.typesetter = oldTypesetter
+  SILE.settings.popState()
+  -- Important, remove the frame from SILE.frames (it was added there by
+  -- SILE.newFrame()), now that we no longer need it. Otherwise, the
+  -- performances get awful as all our small frames are kept and solved!
+  SILE.frames[parboxFrame.id] = nil
+  return innerVbox
+end
+
+-- PARBOXING COMMAND
+
+local drawBorders = function (x, y, w, h, border, bordercolor)
+  -- The border was initially a debug feature, but it turned out to be neat
+  -- for tables (e.g. the ptable package).
+  -- There's a little ugly tweak here, the bottom and right borders are drawn
+  -- "outside" the box, so that successive parboxes have overlapping borders.
+  -- Tables (ptable package) rely on it... That's not perfect, but might not be
+  -- too much noticeable with a normal border thickness below 1pt or so...
+  if bordercolor then SILE.outputter:pushColor(bordercolor) end
+  if border[1] > 0 then SILE.outputter:drawRule(x, y, w, border[1]) end
+  if border[2] > 0 then SILE.outputter:drawRule(x, y + h, w, border[2]) end
+  if border[3] > 0 then SILE.outputter:drawRule(x, y, border[3], h + border[2]) end
+  if border[4] > 0 then SILE.outputter:drawRule(x + w, y, border[4], h + border[2]) end
+  if bordercolor then SILE.outputter:popColor() end
+end
+
+local insertStruts = function (vboxlist, strut)
+  -- The core assumption here is that first/last vboxes are actual text
+  -- lines. Could be wrong...
+  local h, d = SU.cast("length", strut.height), SU.cast("length", strut.depth)
+  for i = 1, #vboxlist do
+    if vboxlist[i].is_vbox and vboxlist[i].height < h then
+      vboxlist[i].height = h -- Hack height of the first vbox
+      break
+    end
+  end
+  for i = #vboxlist, 1, - 1 do
+    if vboxlist[i].is_vbox and vboxlist[i].depth < d then
+      vboxlist[i].depth = d -- Hack depth of the last vbox
+      break
+    end
+  end
+end
+
+local parseBorderOrPadding = function (rawspec, opt)
+  local spec
+  if type(rawspec) == "table" then
+    spec = rawspec
+  else
+    spec = {}
+    for token in SU.gtoke(rawspec, "[ ]+") do
+      if(token.string) then
+        local value = SU.cast("measurement", token.string)
+        spec[#spec+1] = value:tonumber()
+      end
+    end
+  end
+  if #spec == 1 then
+    return { spec[1], spec[1], spec[1], spec[1] }
+  end
+  if #spec ~= 4 then SU.error("Invalid "..opt.." specification: "..rawspec) end
+  return spec
+end
+
+local computeNaturalWidth = function (nodes)
+  -- Derived from typesetter:computeLineRatio()
```

Comment:
Which does plenty of wrong things (#1632), so this is gonna be wrong too.

Reply:
Reimplemented based on `typesetter::computeLineRatio()` also returning the natural totals since  PR #1378